File: src/commands/diff.ts
import { diff } from "@libs/diff";
import { exists, walk } from "@std/fs";
import type { Logger } from "../core/logger.ts";
import { blue, green, red } from "@std/fmt/colors";

type Args = {
  fromDir: string;
  toDir: string;
  extra?: {
    logger?: Logger;
    diff?: {
      color?: boolean;
    };
    color?: boolean;
  };
};

export default async ({ fromDir, toDir, extra = {} }: Args) => {
  const { logger, diff: diffOptions, color } = extra;

  let hasFilesWithChanges = false;

  for await (
    const file of walk(fromDir, { includeDirs: false, includeSymlinks: false })
  ) {
    if (!file.isFile) continue;

    const from = file.path;
    const to = from.replace(fromDir, toDir);

    if (!await exists(from, { isFile: true, isReadable: true })) {
      await logger?.log.warn(
        `File ${color ? blue(from) : from} is not readable, skipping...`,
      );

      continue;
    }

    if (!await exists(to, { isFile: true, isReadable: true })) {
      await logger?.log.warn(
        `File ${
          color ? blue(to) : to
        } does not exists or is not a file or is not readable, skipping...`,
      );

      continue;
    }

    await logger?.logNoNl.log(
      `Diffing ${color ? blue(from) : from} against ${
        color ? blue(to) : to
      } ...`,
    );

    const [fc1, fc2] = await Promise.all([
      Deno.readTextFile(from),
      Deno.readTextFile(to),
    ]);

    const diffStr = diff(fc1, fc2, { colors: diffOptions?.color ?? false });

    if (diffStr.length > 0) {
      await logger?.log.log(color ? red(" ✕") : " ✕");
      await logger?.log.log(diffStr);

      hasFilesWithChanges = true;
    } else {
      await logger?.log.log(color ? green(" ✓") : " ✓");
    }
  }

  return !hasFilesWithChanges;
};

File: src/commands/diff.go
package commands

import (
	"fmt"
	"io/fs"
	"path/filepath"
	"strings"

	"github.com/aymanbagabas/go-udiff"
	"github.com/fatih/color"
	"github.com/gookit/goutil/fsutil"
	"github.com/m4rc3l05/dots/src/core"
)

type DiffArgs struct {
	FromDir string
	ToDir   string
}

func (c Commands) Diff(args DiffArgs) (bool, error) {
	hasFilesWithChanges := false

	fromFormatted, err := filepath.Abs(args.FromDir)
	if err != nil {
		return false, err
	}

	args.FromDir = fromFormatted

	toFormatted, err := filepath.Abs(args.ToDir)
	if err != nil {
		return false, err
	}

	args.ToDir = toFormatted

	if !fsutil.IsDir(args.FromDir) || !core.IsPathReadable(args.FromDir) {
		return false, fmt.Errorf(
			"path %s does not exists or is not a directory or is not readable",
			args.FromDir,
		)
	}

	if !fsutil.IsDir(args.ToDir) || !core.IsPathReadable(args.ToDir) {
		return false, fmt.Errorf(
			"path %s does not exists or is not a directory or is not readable",
			args.ToDir,
		)
	}

	err = filepath.WalkDir(args.FromDir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}

		if !d.Type().IsRegular() {
			return nil
		}

		from := path
		to := strings.Replace(from, args.FromDir, args.ToDir, 1)

		if !fsutil.FileExist(from) || !core.IsPathReadable(from) {
			c.Logger.Warnnl(
				"File %s does not exists or is not a file or is not readable, skipping...",
				color.BlueString(from),
			)

			return nil
		}

		if !fsutil.FileExist(to) || !core.IsPathReadable(to) {
			c.Logger.Warnnl(
				"File %s does not exists or is not a file or is not readable, skipping...",
				color.BlueString(to),
			)

			return nil
		}

		c.Logger.Log(
			"Diffing %s against %s ...",
			color.BlueString(from),
			color.BlueString(to),
		)

		diffs := udiff.Unified(from, to, string(fsutil.ReadFile(from)), string(fsutil.ReadFile(to)))

		if len(diffs) > 0 {
			hasFilesWithChanges = true

			c.Logger.Lognl(color.RedString(" ✕"))

			for line := range strings.SplitSeq(strings.TrimSpace(diffs), "\n") {
				if strings.HasPrefix(line, "@@") && strings.HasSuffix(line, "@@") {
					c.Logger.Lognl(color.CyanString(line))
				} else if len(line) > 1 && line[0] == '+' && line[1] != '+' {
					c.Logger.Lognl(color.GreenString(line))
				} else if len(line) > 1 && line[0] == '-' && line[1] != '-' {
					c.Logger.Lognl(color.RedString(line))
				} else {
					c.Logger.Lognl(line)
				}
			}
		} else {
			c.Logger.Lognl(color.GreenString(" ✓"))
		}

		return nil
	})

	return err == nil && !hasFilesWithChanges, err
}
